From: Venkatesh Pallipadi <venki@google.com>
Subject: x86: Look for IA32_ENERGY_PERF_BIAS support
References: bnc#702604
Patch-Mainline: yes
Git-commit: 23016bf0d25d62c45d8b8f61d55b290d704f7a79

Signed-off-by: Thomas Renninger <trenn@suse.de>

The new IA32_ENERGY_PERF_BIAS MSR allows system software to give
hardware a hint whether OS policy favors more power saving,
or more performance.  This allows the OS to have some influence
on internal hardware power/performance tradeoffs where the OS
has previously had no influence.

The support for this feature is indicated by CPUID.06H.ECX.bit3,
as documented in the Intel Architectures Software Developer's Manual.

This patch discovers support of this feature and displays it
as "epb" in /proc/cpuinfo.

Signed-off-by: Venkatesh Pallipadi <venki@google.com>
LKML-Reference: <alpine.LFD.2.00.1006032310160.6669@localhost.localdomain>
Signed-off-by: Len Brown <len.brown@intel.com>
Signed-off-by: H. Peter Anvin <hpa@linux.intel.com>

---
 arch/x86/include/asm/cpufeature.h          |    1 +
 arch/x86/include/asm/msr-index.h           |    2 ++
 arch/x86/kernel/cpu/addon_cpuid_features.c |    3 ++-
 3 files changed, 5 insertions(+), 1 deletion(-)

Index: linux-2.6.32-SLE11-SP2/arch/x86/include/asm/cpufeature.h
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/arch/x86/include/asm/cpufeature.h
+++ linux-2.6.32-SLE11-SP2/arch/x86/include/asm/cpufeature.h
@@ -161,6 +161,7 @@
  */
 #define X86_FEATURE_IDA		(7*32+ 0) /* Intel Dynamic Acceleration */
 #define X86_FEATURE_ARAT	(7*32+ 1) /* Always Running APIC Timer */
+#define X86_FEATURE_EPB		(7*32+ 3) /* IA32_ENERGY_PERF_BIAS support */
 #define X86_FEATURE_XSAVEOPT	(7*32+ 4) /* "xsaveopt" Optimized Xsave */
 
 /* Virtualization flags: Linux defined, word 8 */
Index: linux-2.6.32-SLE11-SP2/arch/x86/include/asm/msr-index.h
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/arch/x86/include/asm/msr-index.h
+++ linux-2.6.32-SLE11-SP2/arch/x86/include/asm/msr-index.h
@@ -237,6 +237,8 @@
 
 #define MSR_IA32_MISC_ENABLE		0x000001a0
 
+#define MSR_IA32_ENERGY_PERF_BIAS	0x000001b0
+
 /* MISC_ENABLE bits: architectural */
 #define MSR_IA32_MISC_ENABLE_FAST_STRING	(1ULL << 0)
 #define MSR_IA32_MISC_ENABLE_TCC		(1ULL << 1)
Index: linux-2.6.32-SLE11-SP2/arch/x86/kernel/cpu/addon_cpuid_features.c
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/arch/x86/kernel/cpu/addon_cpuid_features.c
+++ linux-2.6.32-SLE11-SP2/arch/x86/kernel/cpu/addon_cpuid_features.c
@@ -33,7 +33,8 @@ void __cpuinit init_scattered_cpuid_feat
 	static const struct cpuid_bit __cpuinitconst cpuid_bits[] = {
 		{ X86_FEATURE_IDA,		CR_EAX, 1, 0x00000006, 0 },
 		{ X86_FEATURE_ARAT,		CR_EAX, 2, 0x00000006, 0 },
-		{ X86_FEATURE_XSAVEOPT,		CR_EAX,	0, 0x0000000d, 1 },
+		{ X86_FEATURE_EPB,		CR_ECX, 3, 0x00000006, 0 },
+ 		{ X86_FEATURE_XSAVEOPT,		CR_EAX,	0, 0x0000000d, 1 },
 		{ 0, 0, 0, 0, 0 }
 	};
 
